subscription/src/main/java/org/killbill/billing/subscription/api/user/DefaultSubscriptionBaseApiService.java

```diff
@@ -257,7 +245,22 @@ private boolean doCancelPlan(final Map<DefaultSubscriptionBase, DateTime> subscr
 
         try {
             for (final DefaultSubscriptionBase subscription : subscriptions.keySet()) {
+
+                final EntitlementState currentState = subscription.getState();
+                if (currentState == EntitlementState.CANCELLED) {
+                    throw new SubscriptionBaseApiException(ErrorCode.SUB_CANCEL_BAD_STATE, subscription.getId(), currentState);
+                }
                 final DateTime effectiveDate = subscriptions.get(subscription);
+
+                // If subscription was future cancelled at an earlier date we disallow the operation -- i.e,
+                // user should first uncancel prior trying to cancel again
+                final SubscriptionBaseTransition pendingTransition = subscription.getPendingTransition();
+                if (pendingTransition != null &&
+                    pendingTransition.getTransitionType() == SubscriptionBaseTransitionType.CANCEL &&
+                    pendingTransition.getEffectiveTransitionTime().compareTo(effectiveDate) < 0) {
```

Comment:
Aren't we guaranteed that `getPendingTransition` returns future transitions only (`FUTURE_ONLY`)? Otherwise, if the check is really needed, do we need `<=` instead?

Reply:
Updated comment in f7755f9fbb1913efff54a3ae9f4f2326897b295f